src/v/iceberg/tests/metadata_query_test.cc

```diff
@@ -0,0 +1,414 @@
+    auto [snapshots, files, manifests] = execute_for_all_results(tx.table());
+    ASSERT_EQ(snapshots.size(), 5);
+    // all snapshots should be present in the result
+    ASSERT_EQ(snapshots, tx.table().snapshots);
+
+    // all manifest_files should be present in result
+    ASSERT_EQ(files, collect_all_manifest_files(tx.table()).get());
+
+    // all manifests should be present in result
+    auto all = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests, collect_all_manifests(tx.table()).get());
+}
+
+bool is_even(int64_t v) { return v % 2 == 0; }
+
+MATCHER_P(EqRef, expected_value, "") { return arg.value == expected_value; }
+TEST_F(MetadataQueryTest, TestQuerySnapshots) {
+    auto tx = generate_metadata().get();
+    // snapshot query, only those snapshots which has even sequence number,
+    // those will be snapshot 2 and 4
+    auto s_matcher = [](const snapshot& s) {
+        return is_even(s.sequence_number());
+    };
+
+    auto [snapshots, files, manifests] = execute_for_all_results(
+      tx.table(), s_matcher);
+    ASSERT_EQ(snapshots.size(), 2);
+    EXPECT_THAT(
+      snapshots, Each(Field(&snapshot::sequence_number, Truly(is_even))));
+
+    auto all_manifest_files = collect_all_manifest_files(tx.table()).get();
+
+    /**
+     * The ordering here is counter intuitive.
+     * Table metadata contains 5 snapshots [s1,s2,s3,s4,s5] each snapshot
+     * contains metadata files
+     *
+     * s1 - [f0]
+     * s2 - [f1,f0]
+     * s3 - [f2,f1,f0]
+     * s4 - [f3,f2,f1,f0]
+     * s5 - [f4,f3,f2,f1,f0]
+     *
+     * When test util function collects all files/manifests it traverses the
+     * snapshot list and deduplicate the files. The result of this operation is:
+     * [f0,f1,f2,f3,f4]
+     *
+     * When query is executed with is even segment sequence it matches s2 and s4
+     * from s2 it takes
+     *
+     * [f1, f0] and then from s4 [f3,f2]
+     *
+     * So the end result is [f1,f0,f3,f2].
+     */
+    ASSERT_EQ(files.size(), 4);
+    ASSERT_EQ(files[0], all_manifest_files[1]);
+    ASSERT_EQ(files[1], all_manifest_files[0]);
+    ASSERT_EQ(files[2], all_manifest_files[3]);
+    ASSERT_EQ(files[3], all_manifest_files[2]);
+
+    // the same ordering rule applies to the manifests
+    auto all_manifests = collect_all_manifests(tx.table()).get();
+    ASSERT_EQ(manifests.size(), 4);
+    ASSERT_EQ(manifests[0], all_manifests[1]);
+    ASSERT_EQ(manifests[1], all_manifests[0]);
+    ASSERT_EQ(manifests[2], all_manifests[3]);
+    ASSERT_EQ(manifests[3], all_manifests[2]);
+}
+
+TEST_F(MetadataQueryTest, TestQueryManifestFile) {
+    auto tx = generate_metadata().get();
+    // manifest file query, only those manifest files which partition key
+    // lower bound is greater or equal than 3
+    auto matcher = [](const manifest_file& file) {
+        return *(reinterpret_cast<const int*>(
+                 file.partitions[0].lower_bound->cbegin()))
+               >= 3;
```

Comment:
nit: maybe use `std::get<int_value>(value_from_bytes(...)`? Would be a bit clearer that it's evaluating the right value